test/connection_test.exs

```diff
@@ -11,4 +41,164 @@ defmodule Janus.ConnectionTest do
+      response = %{"janus" => "ack"}
+      message = Map.merge(response, %{"transaction" => transaction})
+
+      Connection.handle_info(
+        message,
+        state(
+          pending_calls_table: table,
+          transport_module: ValidTransport
+        )
+      )
+
+      assert_receive response, @receive_timeout
+
+      assert {:error, :unknown_transaction} = Transaction.transaction_status(table, transaction)
+    end
+
+    test "not respond to expired transaction", %{table: table} do
+      past = @now |> DateTime.add(-20_000, :second)
+      transaction = Transaction.insert_transaction(table, from(), 0, past)
+      response = %{"janus" => "ack"}
+      message = Map.merge(response, %{"transaction" => transaction})
+
+      Connection.handle_info(
+        message,
+        state(
+          pending_calls_table: table,
+          transport_module: ValidTransport
+        )
+      )
+
+      refute_receive response, @receive_refute
+      assert {:error, :unknown_transaction} = Transaction.transaction_status(table, transaction)
+    end
+
+    test "batched response", %{table: table} do
+      batch_size = 5
+      timeout = 20_000_000
+      janus_response = %{"janus" => "success"}
+
+      transactions =
+        for _ <- 1..batch_size, do: Transaction.insert_transaction(table, from(), timeout)
+
+      messages =
+        for t <- transactions, do: Map.merge(janus_response, %{"data" => t, "transaction" => t})
+
+      Connection.handle_info(
+        messages,
+        state(
+          pending_calls_table: table,
+          transport_module: ValidTransport
+        )
+      )
+
+      for t <- transactions, do: assert_next_receive(t, @receive_timeout)
+    end
+  end
+
+  describe "init should" do
+    test "fail if transport's connect fails" do
+      assert {:stop, {:connect, "transport"}} =
+               Connection.init({BrokenTransport, [], ValidHandler, []})
+    end
+
+    test "fail if handler's init fails" do
+      assert {:stop, {:handler, "handler"}} =
+               Connection.init({ValidTransport, [], BrokenHandler, []})
+    end
+
+    test "initialize state" do
+      cleanup_timeout = 0
+
+      assert {:ok,
+              state(
+                transport_module: ValidTransport,
+                transport_state: "transport",
+                handler_module: ValidHandler,
+                handler_state: "handler",
+                pending_calls_table: table
+              )} = Connection.init({ValidTransport, [], ValidHandler, []}, cleanup_timeout)
+
+      assert_receive :cleanup, @receive_timeout
+      assert table in :ets.all()
+    end
+  end
+
+  describe "call should" do
+    defmodule ValidTransportMock do
+      def send(_, _, _), do: {}
+    end
+
+    test "send transaction through transport and add transaction to transaction store", %{
+      table: table
+    } do
+      with_mock(ValidTransportMock, send: fn _payload, _timeout, _state -> {:ok, "dummy"} end) do
```

Comment:
this is an add hoc mock and module has async: true

Reply:
If you use Transport from master this mock will not be needed I think. I might be wrong though.

Reply:
`assert_called` works only with mocked modules and Mock does it globally. If we want to keep `Janus.ConnectionTest` asynchronous then it is forbidden to mock modules that might be used in other tests.

Reply:
Add TODO to replace this with MockTransport when we add function to assert on length of response queue inside the MockTransport.